### PR TITLE
Add support for installing the CLI on DBR

### PR DESCRIPTION
**Pull request by shreyas-goenka**
We want to install the CLI at $HOME/bin when the `install.sh` installation script is invoked from DBR. This allows us to provide the CLI in DBR though a proxy trampoline script that's installed in DBR at `/usr/local/bin`.

This PR also adds a couple of tests and assertions to make this functionality robust for use in DBR.

### CODE REVIEW

File: install.sh [hunk 1]
```diff
@@ -27,6 +27,11 @@ MINGW64_NT)
     ;;
 esac
 
+# Set target to ~/bin if DATABRICKS_RUNTIME_VERSION environment variable is set.
+if [ -n "$DATABRICKS_RUNTIME_VERSION" ]; then
+    TARGET="$HOME/bin"
```
Comment (pietern) [inline install.sh:32]: Does this path always exist?

Comment (shreyas-goenka) [reply install.sh:32]: No. Added a mkdir command.

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -79,13 +79,62 @@ jobs:
...
+      - name: Assert the version of the CLI installed
+        run: ./setup-cli/assert/version.sh $(cat ./setup-cli/VERSION)
+        shell: bash
+
+      - name: Assert installation path is /usr/local/bin for non-windows platforms
+        run: ./setup-cli/assert/path.sh /usr/local/bin/databricks
+        if: matrix.os != 'windows-latest'
+        shell: bash
+
+      - name: Assert installation path is /c/Windows for windows platforms
```
Comment (andrewnester) [inline .github/workflows/test.yml:101]: Why would it matter?

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -79,13 +79,62 @@ jobs:
...
+
+      - name: Assert databricks CLI is not already installed
+        run: ./setup-cli/assert/not-installed.sh
+        shell: bash
+
+      - name: Install the CLI
+        run: ./setup-cli/install.sh
+        shell: bash
+        env:
+          DATABRICKS_RUNTIME_VERSION: value-does-not-matter
```
[the remark this answers is not in the document]
Comment (shreyas-goenka) [reply .github/workflows/test.yml:126]: While it's not documented to be guaranteed, it's been present on all clusters I have tried out and is well-established as a rubric to detect that code is being run from DBR.

For example, Mlflow also uses the same check here
https://github.com/mlflow/mlflow/blob/30462203357010cba51d743a96c0cae5397e7e1b/mlflow/utils/databricks_utils.py#L180

[the remark this answers is not in the document]
Comment (shreyas-goenka) [reply .github/workflows/test.yml:126]: Or the python SDK:
https://github.com/databricks/databricks-sdk-py/blob/3a2798f608da46ec44eebabae21b398d22ccb8f8/databricks/sdk/credentials_provider.py#L88

File: assert/not-installed.sh [hunk 4]
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env bash
+
+set -euo pipefail
+
+# Assert databricks CLI is not already installed
+if which databricks >/dev/null 2>&1; then
+    exit 1
+else
+    exit 0
+fi
```
Comment (pietern) [inline assert/not-installed.sh:10]: FWIW, this shortens to `! which databricks`.

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -104,11 +153,57 @@ jobs:
...
+  curl-dbr:
+    # All DBR images are built on top of Ubuntu
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          path: ./setup-cli
+
+      # Append ~/bin to the PATH. This helps with the assertions in the next steps.
```
Comment (pietern) [inline .github/workflows/test.yml:188]: How does it help?

E.g. "When run on DBR, the installation scripts installs to ~/bin. Add it to $PATH to make assertions work."

File: .github/workflows/test.yml [hunk 6]
```diff
@@ -104,11 +153,57 @@ jobs:
...
+      - run: echo "PATH=$PATH:$HOME/bin" >> $GITHUB_ENV
```
Comment (pietern) [inline .github/workflows/test.yml:189]: Can use `$GITHUB_PATH`

https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#adding-a-system-path

File: .github/workflows/test.yml [hunk 7]
```diff
@@ -104,11 +153,57 @@ jobs:
...
+        run: ./setup-cli/assert/not-installed.sh
+        shell: bash
+
       - run: curl -fsSL https://raw.githubusercontent.com/databricks/setup-cli/${{ github.sha }}/install.sh | sh
         shell: bash
 
       - run: databricks version
         shell: bash
 
-      - run: ./setup-cli/test.sh $(cat ./setup-cli/VERSION)
```
Comment (pietern) [inline .github/workflows/test.yml:113]: Is `test.sh` still used?

File: .github/workflows/test.yml (hunk 5 above)
Comment (shreyas-goenka) [reply .github/workflows/test.yml:188]: Addressed in https://github.com/databricks/setup-cli/pull/91

File: .github/workflows/test.yml (hunk 6 above)
Comment (shreyas-goenka) [reply .github/workflows/test.yml:189]: addressed in https://github.com/databricks/setup-cli/pull/91

File: assert/not-installed.sh (hunk 4 above)
Comment (shreyas-goenka) [reply assert/not-installed.sh:10]: Yeah. I personally find the current approach clearer and more explicit and consistent with the other assertion definitions though.

**Pull request merged**